File: kickstart/2020/RoundC/perfect_array.py
import math


def check_square(x):
    if x >= 0:
        sq = math.sqrt(x)
        return (sq - math.floor(sq)) == 0
    return False


def cumulative_sum(arr):
    N = len(arr)
    cum_sum = [0] * (N+1)
    for i in range(N):
        cum_sum[i+1] = cum_sum[i] + arr[i]
    return cum_sum


def perfect_array():
    T = int(input())
    for t in range(1, T + 1):
        N = int(input())
        arr = list(map(int, input().split()))
        cum_sum = cumulative_sum(arr)
        res = 0
        for i in range(N):
            for j in range(i, N):
                total = cum_sum[j+1] - cum_sum[i]
                if check_square(total):
                    res += 1
        print('Case #{}: {}'.format(t, res))


def perfect_array2():
    T = int(input())
    for t in range(1, T + 1):
        N = int(input())
        arr = list(map(int, input().split()))
        res = 0
        curr_sum, min_sum = 0, 0
        m = {0: 1}
        for i in range(N):
            curr_sum += arr[i]
            min_sum = min(min_sum, curr_sum)
            j = 0
            while curr_sum - (j * j) >= min_sum:
                if curr_sum - (j * j) in m:
                    res += m[curr_sum - j*j]
                j += 1
            if curr_sum in m:
                m[curr_sum] += 1
            else:
                m[curr_sum] = 1
        print('Case #{}: {}'.format(t, res))


def test_cumulative_sum():
    assert cumulative_sum([1, 2, 3, 4, 5]) == [0, 1, 3, 6, 10, 15]


if __name__ == '__main__':
    perfect_array2()

File: kickstart/2020/RoundC/perfect_array.cpp
#include <bits/stdc++.h>
typedef long long ll;
using namespace std;

bool check_perfect(int sum)
{
    if (sum >= 0)
    {
        float sq = sqrt(float(sum));
        return (sq - floor(sq)) == 0;
    }
    return false;
}

void print_vector(vector<int> &v)
{
    for (int i = 0; i < v.size(); ++i)
    {
        cout << v[i];
        if (i != v.size() - 1)
        {
            cout << ',';
        }
    }
    cout << endl;
}

vector<int> cumulative_sum(vector<int> &arr)
{
    vector<int> cum_sum;
    cum_sum.resize(arr.size() + 1);
    cum_sum[0] = 0;
    for (int i = 0; i < arr.size(); ++i)
    {
        cum_sum[i + 1] = cum_sum[i] + arr[i];
    }
    return cum_sum;
}

int main()
{
    ll T, N;
    cin >> T;
    for (ll t = 1; t <= T; ++t)
    {
        cin >> N;
        vector<ll> arr;
        for (ll i = 0; i < N; ++i)
        {
            int d;
            cin >> d;
            arr.push_back(d);
        }
        ll res = 0;
        ll sum = 0;
        ll min_sum = 0;
        unordered_map<ll, ll> m;
        m[0] = 1;
        for (ll i = 0; i < arr.size(); ++i)
        {
            sum += arr[i];
            min_sum = min(sum, min_sum);
            for (ll j = 0; sum - (j * j) >= min_sum; ++j)
            {
                if (m.count(sum - (j * j)) == 1)
                {
                    res += m[sum - (j * j)];
                }
            }
            m[sum] += 1;
        }
        cout
            << "Case #" << t << ": " << res << endl;
    }
}
